### PR TITLE
Tile documentation

### PR DESCRIPTION
**Pull request by LeoDog896**
Documentation is _very_ important for the modding community, and there will finally be a time in the future where modders don't have to download Mindustry's source code just to understand what's going on there. This just does that, but for tiles.

### CODE REVIEW

**Comment by SnakkiZXZ** [comment]
oh yes this documentation is great

File: core/src/mindustry/world/CachedTile.java [hunk 1]
```diff
@@ -27,14 +27,14 @@ protected void changeEntity(Team team, Prov<Building> entityprov, int rotation){
...
         if(block.hasBuilding()){
-            Building n = entityprov.get();
-            n.cons(new ConsumeModule(build));
-            n.tile(this);
-            n.block(block);
-            if(block.hasItems) n.items = new ItemModule();
-            if(block.hasLiquids) n.liquids(new LiquidModule());
-            if(block.hasPower) n.power(new PowerModule());
-            build = n;
+            Building building = entityprov.get();
```
Comment (Anuken) [inline core/src/mindustry/world/CachedTile.java:30]: Use `var building` here if you're going to use the full class name anyway

File: core/src/mindustry/world/Tile.java [hunk 2]
```diff
@@ -62,7 +62,9 @@ public byte relativeTo(Tile tile){
         return relativeTo(tile.x, tile.y);
     }
 
-    /** Return relative rotation to a coordinate. Returns -1 if the coordinate is not near this tile. */
+    /**
+     * @return relative rotation to a coordinate. Returns -1 if the coordinate is not near this tile.
```
Comment (Anuken) [inline core/src/mindustry/world/Tile.java:66]: Use single-line documentation, e.g. `/** @return xyz */`

File: core/src/mindustry/world/TileGen.java [hunk 3]
```diff
@@ -4,13 +4,16 @@
...
-    public Block block ;
+    public Block block;
     public Block overlay;
 
     {
         reset();
     }
 
+    /**
+     * Resets this TileGen back to its normal state.
```
Comment (Anuken) [inline core/src/mindustry/world/TileGen.java:15]: one line

File: core/src/mindustry/world/Tiles.java [hunk 4]
```diff
@@ -7,18 +7,40 @@
 
 import java.util.*;
 
-/** A tile container. */
+/**
+ * Contains a list of tiles with extended behavior
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:11]: This class should use one line whenever possible, like `Block` does:
```java
/** doc 1 */
int field1;
/** doc 2 */
int field2;
```

Extra newlines waste a lot of space and make things harder to read.

File: core/src/mindustry/world/Tiles.java [hunk 5]
```diff
@@ -27,64 +49,127 @@ public void each(Intc2 cons){
...
+     */
     public @NonNull Tile geti(int idx){
         return array[idx];
     }
 
-    /** @return a tile at an int position (not equivalent to geti) */
+    /**
+     * Get a tile at an int position (not equivalent to geti)
+     *
+     * @param pos The int position
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:138]: If you're going to add docs for this, you should explain what an "int position" is: two short x/y coords packed into an int

File: core/src/mindustry/world/Tiles.java [hunk 6]
```diff
@@ -27,64 +49,127 @@ public void each(Intc2 cons){
...
         }
     }
 
+    /**
+     * iterating through the entire map is expensive anyway, so a new allocation doesn't make much of a difference
+     * @return The TileIterator tied to this tile array.
+     */
     @Override
     public Iterator<Tile> iterator(){
-        //iterating through the entire map is expensive anyway, so a new allocation doesn't make much of a difference
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:83]: this is an implementation comment and should not be part of the doc

File: core/src/mindustry/world/Tiles.java [hunk 7]
```diff
@@ -27,64 +49,127 @@ public void each(Intc2 cons){
         }
     }
 
-    /** fills this tile set with empty air tiles. */
+    /**
+     * Fills this tile set with empty air tiles.
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:53]: lots of wasted space here with a random @

Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:53]: lots of wasted space, extra '@'

File: core/src/mindustry/world/Tiles.java [hunk 8]
```diff
@@ -27,64 +49,127 @@ public void each(Intc2 cons){
...
         return array[idx];
     }
 
-    /** @return a tile at an int position (not equivalent to geti) */
+    /**
+     * Get a tile at an int position (not equivalent to geti)
+     *
+     * @param pos The int position
+     *
+     * @return a tile at an int position
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:140]: explain what an int position is: x/y shorts packed into an int

File: core/src/mindustry/world/Tiles.java (hunk 6 above)
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:83]: implementation detail, should not be part of the doc

File: core/src/mindustry/world/Tiles.java (hunk 4 above)
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:11]: every class should use single-line doc comments whenever possible, multiple blank lines waste a lot of space and don't make anything clearer

File: core/src/mindustry/world/CachedTile.java (hunk 1 above)
Comment (Anuken) [inline core/src/mindustry/world/CachedTile.java:30]: use `var building`

**Review by Anuken — CHANGES_REQUESTED** [review]

File: core/src/mindustry/world/Tiles.java (hunk 7 above)
Comment (LeoDog896) [reply core/src/mindustry/world/Tiles.java]: I honestly don't even know how that got there

File: core/src/mindustry/world/Tiles.java [hunk 9]
```diff
@@ -27,64 +43,119 @@ public void each(Intc2 cons){
...
         for(Tile tile : array){
             cons.get(tile);
         }
     }
 
+    /** @return The TileIterator tied to this tile array. */
     @Override
     public Iterator<Tile> iterator(){
-        //iterating through the entire map is expensive anyway, so a new allocation doesn't make much of a difference
+        // iterating through the entire map is expensive anyway, so a new allocation doesn't make much of a difference
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:149]: unnecessary space

File: core/src/mindustry/world/Tiles.java [hunk 10]
```diff
@@ -27,64 +43,119 @@ public void each(Intc2 cons){
...
         return new TileIterator();
     }
 
+    /**
+     * Implements through tiles with extra power
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:154]: unnecessary multiline + documentation of internal API?

File: core/src/mindustry/world/Tiles.java [hunk 11]
```diff
@@ -7,18 +7,34 @@
 
 import java.util.*;
 
-/** A tile container. */
+/**
+ * Contains a list of tiles with extended behavior
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:11]: unnecessary multiline

File: core/src/mindustry/world/Tiles.java [hunk 12]
```diff
@@ -27,64 +43,119 @@ public void each(Intc2 cons){
...
+        // iterating through the entire map is expensive anyway, so a new allocation doesn't make much of a difference
         return new TileIterator();
     }
 
+    /**
+     * Implements through tiles with extra power
+     */
     private class TileIterator implements Iterator<Tile>{
+
+        /** The index at which the {@link TileIterator} lies. */
```
Comment (Anuken) [inline core/src/mindustry/world/Tiles.java:158]: this isn't public API and it's pretty obvious for anyone reading the source code. I think anyone can tell you that *"field index in class TileIterator"* is *"the index of the TileIterator"*

**Review by Anuken — CHANGES_REQUESTED** [review]

**Comment by Anuken** [comment]
Closing due to stale comments.

**Pull request closed without merging**